### PR TITLE
store/copr: batch replica read

### LINKED ISSUE #42322: Compatible batch copr with load-based replica read
## Enhancement

We have 2 optimizations which improve the performance of distsql read, one is [the batch coprocessor](https://github.com/pingcap/tidb/issues/39361) and another is the [load-based replica read](https://github.com/pingcap/tidb/issues/41664). But they are currently not compatible, because sending batched tasks to the replica store will produce region misses.

https://github.com/pingcap/tidb/blob/fa28db109db7055cead9591cba9c6dcbc48ae009/store/copr/coprocessor.go#L1126-L1132

As an enhancement, the batched task should be re-split before being sent to the replica nodes. It might choose the best replica node in the coprocessor client.

### PR DESCRIPTION
**Pull request by you06**
### What problem does this PR solve?

Issue Number: close #42322

Problem Summary:

Currently, store batched coprocessor is not compatible with load-based replica read.

### What is changed and how it works?

This PR makes them compatible.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: store/copr/coprocessor.go [hunk 1]
```diff
@@ -502,16 +516,25 @@ func (b *batchStoreTaskBuilder) handle(task *copTask) (err error) {
...
 	if err != nil {
 		return err
 	}
 	if batchedTask == nil {
 		return nil
 	}
-	if idx, ok := b.store2Idx[batchedTask.storeID]; !ok || len(b.tasks[idx].batchTaskList) >= b.limit {
+	key := storeReplicaKey{
+		storeID:     batchedTask.storeID,
+		replicaRead: batchedTask.replicaRead,
```
Comment (cfzjywxk) [inline store/copr/coprocessor.go]: It seems this change is not needed now.

File: store/copr/coprocessor.go [hunk 2]
```diff
@@ -1158,6 +1169,11 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
 	if len(worker.req.MatchStoreLabels) > 0 {
 		ops = append(ops, tikv.WithMatchLabels(worker.req.MatchStoreLabels))
 	}
+	if task.redirect2Replica != nil {
+		req.ReplicaRead = true
```
Comment (cfzjywxk) [inline store/copr/coprocessor.go:1173]: Should we move the setting `busyThreshold=0` here so none of the changes would be missing?

File: store/copr/coprocessor.go [hunk 3]
```diff
@@ -502,14 +511,19 @@ func (b *batchStoreTaskBuilder) handle(task *copTask) (err error) {
...
-	batchedTask, err := b.cache.BuildBatchTask(b.bo, task, b.req.ReplicaRead)
+	batchedTask, err := b.cache.BuildBatchTask(b.bo, b.req, task, b.replicaRead)
 	if err != nil {
 		return err
 	}
 	if batchedTask == nil {
 		return nil
 	}
 	if idx, ok := b.store2Idx[batchedTask.storeID]; !ok || len(b.tasks[idx].batchTaskList) >= b.limit {
+		if b.replicaRead == kv.ReplicaReadFollower {
```
Comment (cfzjywxk) [inline store/copr/coprocessor.go:522]: The `replicaRead` is a bit confusing, maybe `loadbasedReplicaRetry` like name is more clear.

File: store/copr/coprocessor.go [hunk 4]
```diff
@@ -339,7 +345,7 @@ func buildCopTasks(bo *Backoffer, ranges *KeyRanges, opt *buildCopTaskOpt) ([]*c
 
 	var builder taskBuilder
 	if req.StoreBatchSize > 0 && hints != nil {
-		builder = newBatchTaskBuilder(bo, req, cache)
+		builder = newBatchTaskBuilder(bo, req, cache, req.ReplicaRead)
```
[the remark this answers is not in the document]
Comment (you06) [reply store/copr/coprocessor.go:348]: We pass followerRead when building replica read tasks meanwhile `req.ReplicaRead == leaderRead`

File: store/copr/coprocessor.go [hunk 5]
```diff
@@ -1158,6 +1175,16 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
...
 		ops = append(ops, tikv.WithMatchLabels(worker.req.MatchStoreLabels))
 	}
+	if task.redirect2Replica != nil {
+		req.ReplicaRead = true
+		req.ReplicaReadType = options.GetTiKVReplicaReadType(kv.ReplicaReadFollower)
+		ops = append(ops, tikv.WithMatchStores([]uint64{*task.redirect2Replica}))
+		if replicaRead == kv.ReplicaReadLeader {
+			replicaRead = kv.ReplicaReadFollower
+			req.ReplicaRead = true
+			req.ReplicaReadType = options.GetTiKVReplicaReadType(replicaRead)
```
Comment (ekexium) [inline store/copr/coprocessor.go:1185]: These 2 lines are duplicated.

File: store/copr/coprocessor.go [hunk 6]
```diff
@@ -1366,18 +1393,21 @@ func (worker *copIteratorWorker) handleBatchRemainsOnErr(bo *Backoffer, rpcCtx *
...
+	busyThresholdFallback := false
 	defer func() {
 		if err != nil {
 			return
 		}
-		worker.storeBatchedNum.Add(uint64(batchedNum - len(remainTasks)))
-		worker.storeBatchedFallbackNum.Add(uint64(len(remainTasks)))
+		if !busyThresholdFallback {
+			worker.storeBatchedNum.Add(uint64(batchedNum - len(remainTasks)))
+			worker.storeBatchedFallbackNum.Add(uint64(len(remainTasks)))
```
Comment (ekexium) [inline store/copr/coprocessor.go:1409]: Is it supposed to be put in an else clause?

File: store/copr/coprocessor.go [hunk 7]
```diff
@@ -1479,6 +1511,19 @@ func (worker *copIteratorWorker) handleBatchCopResponse(bo *Backoffer, rpcCtx *t
 		}
 		appendRemainTasks(t.task)
 	}
+	if regionErr := resp.GetRegionError(); regionErr != nil && regionErr.ServerIsBusy != nil &&
```
Comment (ekexium) [inline store/copr/coprocessor.go:1514]: We might assert `!(busyThresholdFallback && len(bachResps) > 0)`, protecting it from future change.

File: store/copr/coprocessor.go (hunk 6 above)
Comment (you06) [reply store/copr/coprocessor.go:1409]: The counter will be increased when handling the responses of the follower tasks.

File: store/copr/coprocessor.go (hunk 7 above)
Comment (you06) [reply store/copr/coprocessor.go:1509]: I returned an error when this happened.

File: store/copr/coprocessor.go (hunk 5 above)
Comment (you06) [reply store/copr/coprocessor.go]: Good catch, removed.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by ekexium — APPROVED** [review]

**Pull request merged**